File: javascript/coin_change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */

var coinChange = function (coins, amount) {
  const minCoins = new Array(amount + 1).fill(Infinity);
  minCoins[0] = 0;
  coins.forEach(coin => {
    for (i = 1; i < minCoins.length; i++) {
      if (coin <= i) {
        minCoins[i] = Math.min(minCoins[i], minCoins[i - coin] + 1);
      }
    }
  });
  const last = minCoins[minCoins.length - 1];
  return last != Infinity ? last : -1;
};

File: python/coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        minCoins = [float("inf") for amount in range(amount + 1)]
        minCoins[0] = 0
        for coin in coins:
            for amount in range(1, len(minCoins)):
                if coin <= amount:
                    minCoins[amount] = min(minCoins[amount], minCoins[amount - denmon] + 1)
        return minCoins[-1]

coins = [1, 2, 5]
amount = 11
s = Solution()
print(s.coinChange(coins, amount))